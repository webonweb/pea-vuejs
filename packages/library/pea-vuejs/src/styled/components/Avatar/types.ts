export interface AvatarOption {
  src?: string | null
  alt: string
  title?: string
}