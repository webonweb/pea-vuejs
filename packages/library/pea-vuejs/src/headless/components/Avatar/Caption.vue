<template>
  <Primitive :as="as" :as-child="asChild">
    <slot />
  </Primitive>
</template>
<script lang="ts">
import type { PrimitiveProps } from "@pea-vuejs/primitive";
export interface CaptionAvatarProps extends PrimitiveProps {}
</script>

<script lang="ts" setup>
import { Primitive } from "@pea-vuejs/primitive";

withDefaults(defineProps<CaptionAvatarProps>(), {
  as: "figcaption",
});
</script>
