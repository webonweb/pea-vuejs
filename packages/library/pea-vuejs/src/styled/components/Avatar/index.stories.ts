import type { Meta, StoryObj } from '@storybook/vue3';

import { ref, watch } from 'vue';
import hljs from 'highlight.js';
import AvatarComponent from './index.vue'
import CodeEditor from 'simple-code-editor';


const meta: Meta<typeof AvatarComponent> = {
  title: 'Avatar',
  component: AvatarComponent,
  tags: ['autodocs'],
  argTypes: {
    size: { description: 'Parameter to set the size', options: ["small", "medium", "large"] },
    type: { description: 'Parameter to set the type', options: ["square", "circle"] },
    image: { description: 'Parameter to set image', control: "object" },
  },
};

export default meta;

type Story = StoryObj<typeof AvatarComponent>;

export const Avatar: Story = {
  render: (args) => ({
    components: { AvatarComponent, CodeEditor },
    setup() {
      const apiReference = ref()
      const codeEditorRef = ref()
      const propsCode: Record<string, Record<string, string | boolean>[]> = {
        typescript: [
          {
            name: 'size',
            default: 'medium',
            type: 'string',
            description: 'Parameter to set the size',
            required: false,
          },
          {
            name: 'type',
            default: 'circle',
            type: '"square", "circle"',
            description: 'Parameter to set the type',
            required: false,
          },
          {
            name: 'image',
            default: `
            {
              src?: string | null
              alt: string
              title?: string
            }
            `,
            type: 'AvatarOption',
            description: 'Parameter to set image',
            required: true,
          }
        ],
        javascript: [],
      }

      const codeMap: Record<string, string> = {
        typescript: `<template>
  <Avatar :type :size :image />
</template>

<script lang="ts" setup>
import { ref } from 'vue'
import { Avatar } from "pea-vuejs"

const size = ref('medium')
const type = ref('circle')

const image = ref({
  src: ''
  alt: 'Lorem ipsum
  title: 'Lorem ipsum'
})

</script>`,
        javascript: `<template>
  <Avatar.Root>
    <Shared.Image
      :image="{
        ...image,
        placeholder: 'Url to placeolder image',
      }"
    />
    <Avatar.Caption>
      {{ image?.title || image?.alt }}
    </Avatar.Caption>
  </Avatar.Root>
</template>
  
<script lang="ts" setup>
import { ref } from "vue" 
import { Avatar, Shared } from "pea-vuejs/headless";

const image = ref({
  src: ''
  alt: 'Lorem ipsum
  title: 'Lorem ipsum'
})
</script>`
      }

      const instalationMap: Record<string, string> = {
        npm: 'npm install pea-vuejs',
        pnpm: 'pnpm install pea-vuejs',
        yarn: 'yarn add pea-vuejs',
        bun: 'bun add pea-vuejs'
      }

      const code = ref(codeMap.typescript)
      const instalation = ref(instalationMap.npm)

      const handleChangeLang = (lang: string) => {
        codeEditorRef.value.$refs.code.removeAttribute('data-highlighted');
        code.value = codeMap[lang]

        apiReference.value = propsCode[lang].map(item => `
          <tr class="bg-white border-b  hover:bg-gray-50 ">
              <td class="px-6 py-4">
                  ${item.name}
              </td>
              <td class="px-6 py-4">
                ${item.default}
              </td>
              <td class="px-6 py-4">
                ${item.type}
              </td>
              <td class="px-6 py-4">
                ${item.description}
              </td>
              <td class="px-6 py-4">
                ${item.required}
              </td>
          </tr>
        `).join('')
      }

      const handleChangeInstalationLang = (lang: string) => {
        codeEditorRef.value.$refs.code.removeAttribute('data-highlighted');
        instalation.value = instalationMap[lang]
      }


      watch(() => [args.type, args.size, code.value], () => {
        codeEditorRef.value.$refs.code.removeAttribute('data-highlighted');
      })



      return {
        args,
        code,
        apiReference,
        instalation,
        codeEditorRef,
        handleChangeLang,
        handleChangeInstalationLang
      };
    },
    template: `
        <div class="w-[800px] pb-10 overflow-visible">
      <div class="grid grid-rows-[repeat(2,max-content)] gap-y-4">
        <h3 class="text-5xl font-bold">Avatar</h3>
        <p class="text-base text-black/60">The Avatar component in software is used to represent a user through an image, icon, or initials, facilitating identification in user interfaces such as social media applications, user management systems, or messaging platforms. The Avatar often includes customization options, allowing users to change the appearance and tailor the representation to their preferences.</p>
      </div>
      <br/>
      <div class="grid grid-rows-[repeat(2,max-content)] gap-y-4">
        <strong class="text-2xl font-bold"># Example of component usage:</strong>
        <div class="p-8 h-40 bg-black/5 rounded-t-lg border border-black/5 relative flex items-center justify-center">
          <AvatarComponent :image="{ alt: 'Lorem ipsum', title:'title' }" :size="args.size" :type="args.type" caption/>
        </div>
        <CodeEditor ref="codeEditorRef" v-model="code" :read-only="true" :languages="[['typescript', 'Styled'], ['javascript', 'Headless']]" class="!w-[800px] relative -top-[30px]" theme="base16-bespin" @lang="handleChangeLang"/>
      </div>
      <SeparatorComponent /><br/>
      <div class="grid grid-rows-[repeat(2,max-content)] gap-y-4 mb-10">
        <div>
          <strong class="text-2xl font-bold mb-3 block"># Installation:</strong>
          <p class="text-base text-black/60">Install the component from your command line.</p>
        </div>
        <CodeEditor v-model="instalation" :read-only="true" :languages="[['npm', 'npm'], ['pnpm', 'pnpm'], ['yarn', 'yarn'], ['bun', 'bun']]" class="!w-[800px]" theme="base16-bespin" @lang="handleChangeInstalationLang"/>
      </div>
      <SeparatorComponent /><br/>
      <div class="grid grid-rows-[repeat(2,max-content)] gap-y-4">
        <div>
          <strong class="text-2xl font-bold mb-3 block"># API Reference:</strong>
          <p class="text-base text-black/60">This section provides a detailed overview of the properties available in the component. It serves as a guide for developers on how to utilize and configure the component within their applications.</p>
        </div>
        <div>
        <br/>
          <table class="w-full text-sm text-left rtl:text-right text-gray-500 ">
            <thead class="text-xs text-gray-700 bg-gray-100 ">
              <tr>
                <th scope="col" class="px-6 py-3">
                  Prop
                </th>
                <th scope="col" class="px-6 py-3">
                  Default
                </th>
                <th scope="col" class="px-6 py-3">
                  Type
                </th>
                <th scope="col" class="px-6 py-3">
                  Description
                </th>
                <th scope="col" class="px-6 py-3">
                  Required
                </th>
              </tr>
            </thead>
            <tbody v-html="apiReference">
            </tbody>
          </table>
        </div>
      </div>
      <br/>
      <div class="grid grid-rows-[repeat(2,max-content)] gap-y-4">
        <div>
          <strong class="text-2xl mb-3 font-bold block"># Accessibility:</strong>
          <p class="text-base text-black/60">Accessibility for the Avatar component ensures that users with disabilities can interact with and understand the avatar representation through alternative text descriptions and keyboard navigation. This includes providing meaningful alt text for images, ensuring avatars are focusable, and integrating with screen readers to convey necessary information about the user's identity.</p>
        </div>
        <div>
          For more info go to -> <a class=" text-blue-600" href="https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/figure_role" target="__blank">avatar</a> 
        </div>
      </div>
    </div>
  `,
  }),
  args: {
    type: 'square',
    size: 'medium',
  },
};
