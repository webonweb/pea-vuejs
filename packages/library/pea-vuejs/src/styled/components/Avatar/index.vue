<template>
  <Avatar.Root
    class="overflow-hidden border border-pea-surface-black/5"
    :class="schemeClasses"
  >
    <Shared.Image
      :image="{
        ...image,
        placeholder: EmptyAvatar,
      }"
      :class="schemeClasses"
    />
    <Avatar.Caption>
      {{ image?.title || image?.alt }}
    </Avatar.Caption>
  </Avatar.Root>
</template>
6
<script lang="ts" setup>
import { computed } from "vue";
import { Avatar, Shared } from "@/headless";
import type { AvatarOption } from "./types";
import EmptyAvatar from "@/assets/images/images/empty-avatar.webp";

const props = withDefaults(
  defineProps<{
    size?: "small" | "medium" | "large";
    type?: "square" | "circle";
    image: AvatarOption;
  }>(),
  {
    size: "large",
    type: "circle",
  }
);

const variantClassMap: Record<
  "small" | "medium" | "large",
  Record<string, string>
> = {
  small: {
    base: "w-8 h-8",
    circle: "rounded-full",
    square: "",
  },
  medium: {
    base: "w-10 h-10",
    circle: "rounded-full",
    square: "",
  },
  large: {
    base: "w-12 h-12",
    circle: "rounded-full",
    square: "",
  },
};

const schemeClasses = computed(() =>
  [
    variantClassMap[props.size].base,
    variantClassMap[props.size][props.type],
  ].join(" ")
);
</script>
