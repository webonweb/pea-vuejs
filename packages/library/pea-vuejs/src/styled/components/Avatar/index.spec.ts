import { mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';
import Avatar from './index.vue'; // Adjust the import path
import EmptyAvatar from '../../../assets/images/empty-avatar.webp';

describe('Avatar.vue', () => {
  const image = {
    src: 'https://example.com/image.jpg',
    title: 'User Avatar',
    alt: 'User Avatar Alt'
  };


  it('matches the snapshot', () => {
    const wrapper = mount(Avatar, {
      props: { image }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});