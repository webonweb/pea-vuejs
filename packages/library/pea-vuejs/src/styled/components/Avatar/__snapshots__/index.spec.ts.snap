// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar.vue > matches the snapshot 1`] = `
"<figure class="overflow-hidden border border-pea-surface-black/5 w-12 h-12 rounded-full"><img loading="lazy" draggable="false" decoding="async" fetchpriority="auto" src="/src/assets/images/empty-avatar.webp" alt="User Avatar Alt" title="User Avatar" class="w-12 h-12 rounded-full">
  <figcaption>User Avatar</figcaption>
</figure>"
`;
